File: mutator.c
/*---------------------------------------------------------------------
mutator.c - Santiago Garcia -- January 2024 
Description: A basic mutator for .dat files to be loaded into the DSP
----------------------------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>


#define MISSING_ARGUMENTS -1
#define FILE_DOES_NOT_EXIST -2
#define BUF_SIZE 50


int countLines(FILE * filepointer){
    char cur;
    int lines = 1;

    while((cur = getc(filepointer)) != EOF){
        if(cur == ','){
            lines++;
        }
    }
    printf("Number of Lines in the file = %d\n", lines);
    return lines;
}



int main(int argc, char *argv[]){
    
	srand(time(NULL));
	FILE *fp;   
	int lines;
	char number_buf[BUF_SIZE];
	memset(number_buf, '\0', BUF_SIZE);
	

	printf("----------Mutator----------\n");
	if(argc != 2){
		printf("Invalid arguments... try again...\n");
		return(MISSING_ARGUMENTS);
	}
	char * filename = argv[1];

	if(!(fp = fopen(filename,"rb"))){
		printf("Cannot find .dat file specified... try again...\n");
		return(FILE_DOES_NOT_EXIST);
	}
   // printf("Size of Pointer: %lu, Size of File: %lu \n", sizeof(fp), sizeof(*fp));
	
	lines = countLines(fp);
	fp = fopen(filename, "wb");
	rewind(fp);
	int random_num = 0;

	for(int i = 0; i < lines; i++){
    		random_num = rand();

    		if((sprintf(number_buf,"     %d,\n", random_num)) < 0){
        	printf("Error with Random Number Generation\n");
    		}	

    		printf("Adding: %s", number_buf);
	
	//printf("String Length: %ld\n", strlen(number_buf));
		printf("Number of Elements Writen: %ld \n", fwrite(number_buf,sizeof(char), strlen(number_buf), fp));
	}
    



    fclose(fp);
    //TODO 
	// 1. Find better mutations to use for this..
	// 2. 
}

File: mutator.py
import random
import sys

MISSING_ARGUMENTS = -1
FILE_DOES_NOT_EXIST = -2
SUCCESS = 42


def count_lines(file_path):
    with open(file_path, 'r') as file_pointer:
        lines = 0
        for char in file_pointer.read():
            if char == ',':
                lines += 1
        #print("Number of Lines in the file =", lines)
        return lines

def random_one_input(filename):
    random.seed()
    buf_size = 50

    #print("----------Mutator----------")

    try:

        with open(filename, 'wb') as fp:
            random_num = 0

            for i in range(0,1):
                random_num = random.randint(0, 2**31)

                number_str = f"{random_num}"
                print(f"LOG: Trying... {number_str}")

                # Write the string to the file
                fp.write(bytes(number_str, 'utf-8'))
                #print("Number of Elements Written:", fp.write(bytes(number_str, 'utf-8')))
    
    except FileNotFoundError:
        print("Cannot find .dat file specified... try again...")
        return FILE_DOES_NOT_EXIST
    
    return SUCCESS

def random_mutate(filename):
    random.seed()
    buf_size = 50
    number_buf = ['\0'] * buf_size

    #print("----------Mutator----------")

    try:
        with open(filename, 'rb') as fp:
            lines = count_lines(filename)

        with open(filename, 'wb') as fp:
            random_num = 0

            for i in range(lines):
                random_num = random.randint(-2**31, 2**31)

                number_str = f"     {random_num},\n"
                #print(f"Adding: {number_str}")

                # Write the string to the file
                fp.write(bytes(number_str, 'utf-8'))
                #print("Number of Elements Written:", fp.write(bytes(number_str, 'utf-8')))
    
    except FileNotFoundError:
        print("Cannot find .dat file specified... try again...")
        return FILE_DOES_NOT_EXIST
    
    return SUCCESS

def main():
    random.seed()
    buf_size = 50
    number_buf = ['\0'] * buf_size

    #print("----------Mutator----------")
    
    if len(sys.argv) != 2:
        print("Invalid arguments... try again...")
        return MISSING_ARGUMENTS

    filename = sys.argv[1]

    try:
        with open(filename, 'rb') as fp:
            lines = count_lines(filename)

        with open(filename, 'wb') as fp:
            random_num = 0

            for i in range(lines):
                random_num = random.randint(-2**31, 2**31)

                number_str = f"     {random_num},\n"
                print(f"Adding: {number_str}")

                # Write the string to the file
                print("Number of Elements Written:", fp.write(bytes(number_str, 'utf-8')))
    
    except FileNotFoundError:
        print("Cannot find .dat file specified... try again...")
        return FILE_DOES_NOT_EXIST

if __name__ == "__main__":
    main()
